lib/Parse/ParseType.cpp

```diff
@@ -520,7 +520,16 @@ ParserResult<TypeRepr> Parser::parseDeclResultType(Diag<> MessageID) {
     consumeToken(tok::code_complete);
     return makeParserCodeCompletionStatus();
   }
-  return parseType(MessageID);
+
+  auto result = parseType(MessageID);
+
+  if (result.isNonNull() && Tok.is(tok::r_square)) {
+    auto diag = diagnose(Tok, diag::extra_rbracket);
+    diag.fixItInsert(result.get()->getStartLoc(), getTokenText(tok::l_square));
+    consumeToken();
+    return makeParserErrorResult(new (Context) ErrorTypeRepr(Tok.getLoc()));
```

Comment:
Would it be better to diagnose and then form the type we thought the user meant? Or are we worried that we'll pick the wrong alternative in ambiguous cases like `[String: Int]]` and cause further errors?

Reply:
I was following the example of an extra leading `[`, which forms the base type rather than the array type.